MAPPING

artofscience opened the pull request.

artofscience committed:
build the base mapping class, and an exponential and taylor childs

artofscience committed:
change word

artofscience committed:
small changes

artofscience committed:
Merge remote-tracking branch 'origin/main' into combine_approx_interv

artofscience committed:
create a general mapping class and children problem, subproblem and approximation

artofscience committed:
apply feedb ack max

artofscience committed:
changed mapping.py drastically and made a first tested; passed :)

artofscience committed:
some more tests, at some point breaks down (currently commented out)

artofscience committed:
delete subproblem for now (will add later)

artofscience committed:
fixed an error, second order sensitivities not correct yet

artofscience committed:
updated structure:
1. added empty map
2. implemented chain rule

artofscience committed:
Update sao/mappings/mapping.py

Co-authored-by: max <MaxvdKolk@users.noreply.github.com>

artofscience committed:
small change

artofscience committed:
Merge remote-tracking branch 'origin/combine_approx_interv' into combine_approx_interv

artofscience committed:
small change

artofscience committed:
Updated version with approx=intv=mapping

artofscience committed:
Without base class

artofscience committed:
updated emptymap structure

artofscience committed:
remove problem from mapping

artofscience committed:
updat test

artofscience committed:
updat test

artofscience committed:
generated an empty(mapping)

MaxvdKolk commented on sao/mappings/mapping.py:
```diff
@@ -0,0 +1,146 @@
+        return np.zeros_like(x)
+
+
+class Empty(Mapping):
+
+    def __init__(self):
+        pass
+
+    def update(self, x, f, df, ddf=None):
+        pass
+
+    def clip(self, x):
+        return x
+
+    def g(self, x):
+        return x
+
+    def dg(self, x):
+        return np.ones_like(x)
+
+    def ddg(self, x):
+        return np.zeros_like(x)
```
```suggestion
class Empty(Mapping):
    ...
```

It seems behaviour of `Empty()` is the same as the base `Mapping`

artofscience committed:
Apply suggestions from code review

Co-authored-by: max <MaxvdKolk@users.noreply.github.com>

artofscience commented on sao/mappings/mapping.py:
```diff
@@ -0,0 +1,127 @@
+        return np.zeros_like(x)
+
+
+class Empty(Mapping):
+    ...
+
+
```
```suggestion
class Empty(Mapping):
    ...


```

artofscience committed:
Update sao/mappings/mapping.py

artofscience committed:
generate linear mapping as empty mapping

artofscience committed:
small typo

artofscience committed:
fixed error in chain rule, taylor with exp not working yet

artofscience committed:
taylor approximation is working! (does not obey chain rule)

artofscience committed:
test Taylor1(Taylor1(Rec(Lin)))

artofscience committed:
implemented Taylor2 and made drastic simplifications to taylor1/2

artofscience committed:
taylor2 working

artofscience committed:
small improvements

artofscience committed:
moved dx/dy ddx/ddy to mapping update. T2 not working..

artofscience committed:
ARNOUD

artofscience committed:
test approx of approx (aoa)

artofscience committed:
Merge remote-tracking branch 'origin/main' into combine_approx_interv

artofscience committed:
test approx of approx

artofscience committed:
test approx of approx

changed variable names T1 to Linear Approximation and T2 to Diagonal Quadratic Approximation (more accurate description)

artofscience committed:
small modification to aoa:

ddg0 = map.ddg(x)

artofscience committed:
additional tests

artofscience committed:
small mods

artofscience committed:
test_dqa spherical nonspherical
spherical test not working yet!

artofscience committed:
Update pytest.yml

artofscience committed:
remove numba

artofscience committed:
Merge remote-tracking branch 'origin/combine_approx_interv' into combine_approx_interv

artofscience committed:
comment assert

artofscience committed:
Update pytest.yml

artofscience committed:
fix bug spherical

artofscience committed:
added positive/negative and conlin

artofscience committed:
test conlin

artofscience committed:
reformatting structure of files
added mixedmapping class

artofscience committed:
small mod to imports

artofscience committed:
test mapping some initial steps

artofscience committed:
test Mixed Mapping (MM)

artofscience committed:
small mod

artofscience committed:
add mma (not tested yet!)

artofscience committed:
add mma (not tested yet!)

artofscience committed:
modification of exponential initialisation

artofscience committed:
modification of exponential initialisation

dirkmunro89 reviewed (approved):
Had a look. Dont feel I am in a position (yet) to do anything other than approve in principle, and hope to understand better when we discuss.

artofscience committed:
added sum mapping

artofscience committed:
added sum mapping

artofscience committed:
new mapping class (simplicity = key)

artofscience committed:
introduced new and simplified mixed mapping

artofscience committed:
introduced new and simplified mixed mapping

artofscience committed:
introduced new and simplified mixed mapping

artofscience committed:
introduced new and simplified mixed mapping

MaxvdKolk reviewed (commented):
Looks like a nice way to express the classes and the resulting implementations in `change_of_variable.py` are quite compact now. Would `Linear` and `DQA` replace the `Taylor*` classes, or the other way around? 

I wonder if the `eval` function could be done slightly differently, but lets focus on those details later :].

We might also consider to add another function to make the `__setindex__` a bit more discoverable as  `mm[[1], [2]] = Exp(-2)` can be a bit dense. Maybe a wrapper like `mm.set_mapping(response=[1], variables=[2], mapping=Exp(-2))` as well? (which would just forward to `__setindex__` internally)

artofscience committed:
changes based on max's input

artofscience committed:
updated the update method (based on discussion max and stijn) not working yet!

MaxvdKolk commented on sao/mappings/mapping.py:
```diff
@@ -4,35 +4,52 @@
+        self.child.parent = self
+        self.updated = False
+
+    def update(self, x0, dg0, ddg0=0):
+        if self.child is not None and self.child.updated is False:  # climb down to youngest child
+            self.child.update(x0, dg0, ddg0)
+        else:  # climb up to oldest broer while updating the mappings
+            self._update(x0, dg0, ddg0)
+            self.updated = True
+            self.parent.update(self.g(x0), self.dg(x0), self.ddg(x0))
```
I think you might get the behaviour you are looking for by placing the code like this: 

```python
def update(self, x0, dg0, ddg0=0):
    if self.child is not None:
        # Traverse until inner-most child
        x0, dg0, ddg0 = self.child.update(x0, dg0, ddg0)
    
    # Bubble up the results from inside to outside,  and use the
    # updated (x0, dg0, ddg0) tuple from its inner mapping.
    return self._update(x0, dg0, ddg0)

def _update(self, x0, dg0, ddg0):
    # Mapping-specific version of the update rule

artofscience committed:
Commenting the approximations.py file

artofscience committed:
Updated test change of variable

artofscience committed:
Better commenting approximations.py

artofscience committed:
Modified mapping based on maxs comments

artofscience committed:
Some additional comments and setup of test_mapping and test_approx

artofscience committed:
Comment mapping.py
Add tests for mapping

artofscience committed:
Added new test of mapping (update function)

artofscience committed:
Modifed "global" update, such that local "_update" does not require output
